File: dailyprogrammer/304/hard/ifs.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <vector>
#include <fstream>
#include <random>

int RES = 10000;

int WIDTH = 1000;
int HEIGHT = 1000;

bool USE_COLOR = false;

struct color {
	int r,g,b;
	
	color():
		r(0), g(0), b(0)
	{}

	color(int _r, int _g, int _b):
		r(_r), g(_g), b(_b)
	{}
};

struct equ {
	double a,b,c,d,e,f;
	color col;

	double apply_x(double x, double y){
		return a * x + b * y + e;
	}

	double apply_y(double x, double y){
		return c * x + d * y + f;
	}

	std::string to_str(){
		std::stringstream ss;
		ss << a << " " << b << " " << c << " " << d << " " << e << " " << f;
		return ss.str();
	}
};

std::vector<color> COLOR_LIST = {
	color(255,0,0),
	color(255,255,0),
	color(0,255,0),
	color(0,255,255),
	color(0,0,255),
	color(255,0,255)
};

double scale(double x, double in_min, double in_max, double out_min, double out_max){
	return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
}

void write_pbm(const std::vector<std::vector<color>>& img, std::ostream& os){
	os << "P3\n";
	os << img.size() << " " << img[0].size() << "\n";
	os << "255\n";
	for(std::vector<color> v: img){
		for(color& c:v){
			os << c.r << " " << c.g << " " << c.b<< "	";
		}
		os << "\n";
	}
}

int main(int argc, char* argv[]){

	if(argc > 1){
		RES = atoi(argv[1]);
		if(argc > 3){
			WIDTH = atoi(argv[2]);
			HEIGHT = atoi(argv[3]);
			if(argc > 4){
				USE_COLOR = true;
			}
		}
	}

	std::vector<equ> equs;
	std::vector<double> weights;

	std::vector<std::pair<color, std::pair<double, double>>> points;

	while(true){
		equ e;
		double p;
		std::cin >> e.a >> e.b >> e.c >> e.d >> e.e >> e.f >> p;
		if( !e.a && !e.b && !e.c && !e.d && !e.e && !e.f && !p){
			break;
		}
		e.col = COLOR_LIST[equs.size() % COLOR_LIST.size()];
		equs.push_back(e);
		weights.push_back(p);
	}
	//generate random numbers with weights
	std::discrete_distribution<> dist(weights.begin(), weights.end());
	std::random_device rd;
	std::mt19937 gen(rd());

	//initial X and Y
	double x = 1, y = 1;
	double min_x= 100, min_y = 100, max_x = -100, max_y = -100;
	for(int i = 0;i < RES; ++i){
		int n = dist(gen);
		equ e = equs[n];
		//std::cout<<x<<" "<<y<<std::endl;
		if(x < min_x){
			min_x = x;
		}
		if(y < min_y){
			min_y = y;
		}
		if(x > max_x){
			max_x = x;
		}
		if(y > max_y){
			max_y = y;
		}
		points.push_back(std::make_pair(e.col, std::make_pair(x, y)));

		x = e.apply_x(x,y);
		y = e.apply_y(x,y);
	}
	std::vector<std::vector<color>> image(WIDTH + 1, std::vector<color>(HEIGHT + 1, color(255, 255, 255)));

	for(std::pair<color,std::pair<double, double>>& pp: points){
		std::pair<double,double> p = pp.second;
		int image_x = scale(p.first - min_x, 0, max_x - min_x, 0, WIDTH);
		int image_y = scale(p.second - min_y, 0, max_y - min_y, 0, HEIGHT);
		//std::cout<<image_x<<" "<<image_y<<std::endl;

		if(USE_COLOR){
			image[image_x][image_y] = pp.first;
		} else {
			image[image_x][image_y] = color(0,0,0);
		}
	}
	write_pbm(image, std::cout);
}

File: music/ifs/ifs.py
"""Generate a simple IFS image"""

import numpy
from operator import itemgetter

RES = 10000
WIDTH = 1000
HEIGHT = 1000

class equation:
    def equation(self):
        self.a = 0
        self.b = 0
        self.c = 0
        self.d = 0
        self.e = 0
        self.g = 0

    def apply_x(self, x, y):
        return self.a * x + self.b * y + self.e
    def apply_y(self, x, y):
        return self.c * y + self.d * y + self.f

def scale(x, in_min, in_max, out_min, out_max):
    return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min

def write_pbm(image):
    print("P3")
    print("{} {}".format(len(image), len(image[0])))
    print("255")
    for image_row in image:
        s = ""
        for c in image_row:
            s += "{} {} {}\t".format(c[0],c[1],c[2])
        print(s)

def main():
    equations = []
    weights = []
    points = []
    while True:
        line = input().strip()
        lineS = line.split()

        e = equation()
        e.a = float(lineS[0])
        e.b = float(lineS[1])
        e.c = float(lineS[2])
        e.d = float(lineS[3])
        e.e = float(lineS[4])
        e.f = float(lineS[5])
        p = float(lineS[6])

        if p == 0:
            break

        equations.append(e)
        weights.append(p)

    x = 1.0
    y = 1.0

    min_x = 100
    max_x = -100
    min_y = 100
    max_y = -100
    for i in range(RES):
        e = numpy.random.choice(equations, p = weights)
        min_x = min(x, min_x)
        max_x = max(x, max_x)
        min_y = min(y, min_y)
        max_y = max(y, max_y)

        #points.append((int(x*100)/100,int(y*100)/100,))
        points.append((x,y,))
        tx = x
        x = float(e.apply_x(x, y))
        y = float(e.apply_y(tx, y))

    # ps = sorted(points, key=itemgetter(1,0))
    # curX = ps[0][0]
    # s = ""
    # for p in ps:
        # if(p[1] != curX):
            # print(s)
            # s=""
            # curX = p[1]
        # s+="({},{})".format(p[0],p[1])
    
    image = [[(255, 255, 255,)] * (WIDTH+1)] * (HEIGHT+1)
    for point in points:
        image_x = int(scale(point[0] - min_x, 0, max_x - min_x, 0, WIDTH))
        image_y = int(scale(point[1] - min_y, 0, max_y - min_y, 0, HEIGHT))
        #print("({},{})\t({},{})".format(point[0],point[1],image_x,image_y))

        image[image_x][image_y] = (0,0,0,)
    write_pbm(image)

if __name__ == '__main__':
    main()
